virtual_rainforest/models/hydrology/hydrology_model.py

```diff
@@ -702,3 +691,113 @@ def calculate_soil_evaporation(
+    """Find all upstream cell IDs for all grid cells.
+
+    This function identifies all cell IDs that are upstream of each grid cell. This can
+    be used to calculate the water flow that goes though a grid cell.
+
+    Args:
+        lowest_neighbour: list of lowest neighbour cell_ids
+
+    Returns:
+        list of all upstream IDs for each grid cell
+    """
+    upstream_ids: list = [[] for i in range(len(lowest_neighbour))]
+
+    for down_s, up_s in enumerate(lowest_neighbour):
+        upstream_ids[up_s].append(down_s)
+
+    return upstream_ids
+
+
+def accumulate_surface_runoff(
+    drainage_map: dict,
+    surface_runoff: np.ndarray,
+    accumulated_runoff: np.ndarray,
+) -> np.ndarray:
+    """Calculate accumulated surface runoff for each grid cell.
+
+    This function takes the accumulated surface runoff from the previous timestep and
+    adds all surface runoff of the current time step from upstream cell IDs.
+
+    The function currently raises a `ValueError` if accumulated runoff is negative.
+
+    Args:
+        drainage_map: dict of all upstream IDs for each grid cell
+        surface_runoff: surface runoff of the current time step, [mm]
+        accumulated_runoff: accumulated surface runoff from previous time step, [mm]
+
+    Returns:
+        accumulated surface runoff, [mm]
+    """
+
+    upstream_id_list = [[i for i in drainage_map[x]] for x in drainage_map.keys()]
+
+    for cell_id, upstream_id in enumerate(upstream_id_list):
+        accumulated_runoff[cell_id] += np.sum(surface_runoff[upstream_id])
+
+    for num in accumulated_runoff:
+        if num < 0:
+            to_raise = ValueError(
+                "The accumulated surface runoff should not be negative!"
+            )
+            LOGGER.error(to_raise)
+            raise to_raise
+
+    return accumulated_runoff
+
+
+# TODO move this to core.grid once we decided on common use
+def calculate_drainage_map(grid: Grid, elevation: np.ndarray) -> dict[int, tuple[int]]:
+    """Calculate drainage map based on digital elevation model.
+
+    This function finds the lowest neighbour for eacj grid cell, identifies all upstream
+    IDs and creates a dictionnary that provides all upstream cell IDs for each grid cell
+    . This function currently supports only square grids.
+
+    Args:
+        Grid: grid object
+        elevation: elevation, [m]
+
+    Returns:
+        tuple of cell IDs and their upstream neighbours
+    """
+
+    if grid.grid_type != "square":
+        to_raise = ValueError("This grid type is currently not supported!")
+        LOGGER.error(to_raise)
+        raise to_raise
+
+    grid.set_neighbours(distance=sqrt(grid.cell_area))
+    lowest_neighbours = find_lowest_neighbour(grid.neighbours, elevation)
+    upstream_ids = find_upstream_cells(lowest_neighbours)
+
+    output = {}
+    for cell_id in range(grid.n_cells):
+        output[cell_id] = upstream_ids[cell_id]
+
+    return output
```

Comment:
Is there a reason output is a `dict` rather than a `list`?

Reply:
I initially used lists but generated a dict based on @davidorme 's  suggestion. I think it makes sense for easier selection, but I couldn't find a good solution for the internal functions. The suggestions you made to reduce the lists don't work unfortunately

Reply:
It's just that atm you start with `upstream_ids`, which is a `list`, then you create a `dict`, which just has the numbers 0 to `n` as keys (i.e. the same as a `list` would). So I'm not clear on what you get by converting it to a `dict`. Presumably when you extend this function in future you'll sometimes be returning `dict`s without all of these keys?

If you really do want to convert your `list` to a `dict` you can do:

```suggestion
    return dict(enumerate(upstream_ids))
```

(I'm assuming that `grid.n_cells == len(upstream_ids)` here.)

Reply:
@alexdewar the intention was to move this to `core.grid` and have a `dict` that holds all upstream ids with they `cell_id` as key. However, I agree that if it is only me using this here, it is unnecessary to do this and have all these conversions in the middle. Like I said in the previous comment, there is still some clarity needed regarding the types, so I'll sort this out 👍